Sporadic disk cache test failures on 0.54.0dev0 mainline

A couple of caching tests have recently failed on PR builds that do not touch code associated with the tests, only obvious common factors are "disk cache" and `typeguard`. 

1. https://github.com/numba/numba/pull/7064 @ 33c559b failed in `Linux py38_np119_typeguard` with:
<details>

```
9421 tests found. 9421 selected

# Now run the Numba test suite with slicing
# Note that coverage is run from the checkout dir to match the "source"
# directive in .coveragerc
echo "INFO: Running slice of discovered tests: ($TEST_START_INDEX,None,$TEST_COUNT)"
if [ "$RUN_COVERAGE" == "yes" ]; then
    export PYTHONPATH=.
    coverage erase
    $SEGVCATCH coverage run runtests.py -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
elif [ "$RUN_TYPEGUARD" == "yes" ]; then
    echo "INFO: Running with typeguard"
    NUMBA_USE_TYPEGUARD=1 NUMBA_ENABLE_CUDASIM=1 PYTHONWARNINGS="ignore:::typeguard" $SEGVCATCH python runtests.py -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
else
    NUMBA_ENABLE_CUDASIM=1 $SEGVCATCH python -m numba.runtests -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
fi
INFO: Running slice of discovered tests: (13,None,19)
INFO: Running with typeguard
skipped HSA tests
skipped HSA tests
Parallel: 475. Serial: 76
..OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
.......................................................................................................................................s.........................................OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
.....................ssssssssssssssssss.s.................................................................................................................................................................................................................................................................ss...ss............sssss.ss..s..ss..s...sss.......F
Stderr:

-------------------------------------STDERR-------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/support.py", line 861, in _remote_runner
    fn()
  File "/home/vsts/work/1/s/numba/tests/test_extending.py", line 1842, in check_objmode_cache_ndarray_check_cache
    raise AssertionError('unexpected cache miss')
AssertionError: unexpected cache miss

..OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
...........ssssssssss
======================================================================
FAIL: test_check_objmode_cache_ndarray (numba.tests.test_extending.TestCachingOverloadObjmode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_extending.py", line 1857, in test_check_objmode_cache_ndarray
    self.assertEqual(res['exitcode'], 0)
AssertionError: 1 != 0

Stderr:

-------------------------------------STDERR-------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/support.py", line 861, in _remote_runner
    fn()
  File "/home/vsts/work/1/s/numba/tests/test_extending.py", line 1842, in check_objmode_cache_ndarray_check_cache
    raise AssertionError('unexpected cache miss')
AssertionError: unexpected cache miss


----------------------------------------------------------------------
Ran 551 tests in 167.298s
```

</details>

2. https://github.com/numba/numba/pull/7116 @ bf19801 failed in  `Linux py38_np119_typeguard` with:

<details>

```
2021-06-15T15:47:12.6360927Z 9415 tests found. 9415 selected
2021-06-15T15:47:12.6361224Z 
2021-06-15T15:47:12.6361597Z # Now run the Numba test suite with slicing
2021-06-15T15:47:12.6362092Z # Note that coverage is run from the checkout dir to match the "source"
2021-06-15T15:47:12.6362545Z # directive in .coveragerc
2021-06-15T15:47:12.6363062Z echo "INFO: Running slice of discovered tests: ($TEST_START_INDEX,None,$TEST_COUNT)"
2021-06-15T15:47:12.6363576Z if [ "$RUN_COVERAGE" == "yes" ]; then
2021-06-15T15:47:12.6364005Z     export PYTHONPATH=.
2021-06-15T15:47:12.6364387Z     coverage erase
2021-06-15T15:47:12.6365781Z     $SEGVCATCH coverage run runtests.py -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
2021-06-15T15:47:12.6366660Z elif [ "$RUN_TYPEGUARD" == "yes" ]; then
2021-06-15T15:47:12.6367137Z     echo "INFO: Running with typeguard"
2021-06-15T15:47:12.6368375Z     NUMBA_USE_TYPEGUARD=1 NUMBA_ENABLE_CUDASIM=1 PYTHONWARNINGS="ignore:::typeguard" $SEGVCATCH python runtests.py -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
2021-06-15T15:47:12.6369185Z else
2021-06-15T15:47:12.6370124Z     NUMBA_ENABLE_CUDASIM=1 $SEGVCATCH python -m numba.runtests -b -j "$TEST_START_INDEX,None,$TEST_COUNT" --exclude-tags='long_running' -m $TEST_NPROCS -- numba.tests
2021-06-15T15:47:12.6370789Z fi
2021-06-15T15:47:12.6371246Z INFO: Running slice of discovered tests: (13,None,19)
2021-06-15T15:47:12.6371734Z INFO: Running with typeguard
2021-06-15T15:47:21.5192465Z skipped HSA tests
2021-06-15T15:47:21.5193251Z skipped HSA tests
2021-06-15T15:47:21.5193765Z Parallel: 470. Serial: 80
2021-06-15T15:48:33.2613584Z .......................................................................................................................................................................................OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
2021-06-15T15:48:34.1328216Z ......OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
2021-06-15T15:49:03.8587828Z ...........ssssssssssssssssss................s.............................................................................................................................................................................................................................................ss...ss............sssss.ss..s..ss..s...sss..OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
2021-06-15T15:49:20.9047131Z .....Fs...............sssssssssss
2021-06-15T15:49:20.9048615Z ======================================================================
2021-06-15T15:49:20.9049589Z FAIL: test_no_collision (numba.tests.test_dispatcher.TestCacheFileCollision)
2021-06-15T15:49:20.9051190Z ----------------------------------------------------------------------
2021-06-15T15:49:20.9052079Z Traceback (most recent call last):
2021-06-15T15:49:20.9053521Z   File "/home/vsts/work/1/s/numba/tests/test_dispatcher.py", line 1664, in test_no_collision
2021-06-15T15:49:20.9054829Z     self.assertEqual(cachelog.count('index saved'), 1)
2021-06-15T15:49:20.9055640Z AssertionError: 0 != 1
2021-06-15T15:49:20.9056146Z 
2021-06-15T15:49:20.9057183Z ----------------------------------------------------------------------
2021-06-15T15:49:20.9058329Z Ran 550 tests in 119.389s
2021-06-15T15:49:20.9058826Z 
2021-06-15T15:49:20.9059492Z FAILED (failures=1, skipped=49)
2021-06-15T15:49:21.5614825Z ##[error]Bash exited with code '1'.
2021-06-15T15:49:21.5631841Z ##[section]Finishing: Test
```

</details>

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -26,7 +26,7 @@ from numba.np.numpy_support import as_dtype
 from numba.core.caching import _UserWideCacheLocator
 from numba.core.dispatcher import Dispatcher
 from numba.tests.support import (skip_parfors_unsupported, needs_lapack,
-                                 SerialMixin)
+                                 SerialMixin, skip_if_typeguard)
 from numba.testing.main import _TIMEOUT as _RUNNER_TIMEOUT
 import llvmlite.binding as ll
 import unittest
@@ -1598,6 +1598,7 @@ class TestMultiprocessCache(BaseCacheTest):
         self.assertEqual(res, n * (n - 1) // 2)
 
 
+@skip_if_typeguard
 class TestCacheFileCollision(unittest.TestCase):
     _numba_parallel_test_ = False
 

--- a/numba/tests/test_extending.py
+++ b/numba/tests/test_extending.py
@@ -21,6 +21,7 @@ from numba.tests.support import (
     temp_directory,
     override_config,
     run_in_new_process_in_cache_dir,
+    skip_if_typeguard,
 )
 from numba.core.errors import LoweringError
 import unittest
@@ -1162,6 +1163,7 @@ def _assert_cache_stats(cfunc, expect_hit, expect_misses):
         raise AssertionError("cache not used")
 
 
+@skip_if_typeguard
 class TestOverloadMethodCaching(TestCase):
     # Nested multiprocessing.Pool raises AssertionError:
     # "daemonic processes are not allowed to have children"
@@ -1759,6 +1761,7 @@ def with_objmode_cache_ov_example(x):
     pass
 
 
+@skip_if_typeguard
 class TestCachingOverloadObjmode(TestCase):
     """Test caching of the use of overload implementations that use
     `with objmode`

